Grouped Notifications UI

renchap opened the pull request:
This is a big work in progress.

This PR is based on the yet unmerged #29889 PR which implements the backend.

It does not remove the current notifications implementation, but adds a new state/actions/route which uses the new API.

- [x] Typescript types for the API
- [x] Basic Redux state & actions
- [x] Basic `/notifications_v2` route using the new sub-state (100% Typescript)
- [x] Pagination
- [x] Load more (gaps)
- [x] Slow mode (`pending`)
- [x] Streaming integration
- [x] Notifications UI components
- [x] Follow request buttons
- [x] Unread notifications

Target design:
![image-3](https://github.com/mastodon/mastodon/assets/42070/d0096e77-9451-4157-a9a6-eb3feaadde17)

renchap committed:
Implement Redux state for the new grouped notifications API

This is a big WIP

renchap committed:
Add a new experimental notifications route

Gargron committed:
Implement notification components

renchap committed:
Fix pass at fixing lint issues

Gargron committed:
Add unread style

Gargron committed:
Change filtered notifications banner to align with grouped notifications

Gargron committed:
Add click handlers for status previews

renchap committed:
Fix the remaining lint issues

renchap committed:
Fix missing notification types

renchap committed:
Forgot to dispatch the action

renchap committed:
Handle filters, and add basic pagination management

renchap committed:
Add new attribute from #30707

renchap committed:
Change `<LoadGap>` to accept an arbitrary parameter

renchap committed:
Better handle loading notifications from a Gap

Gargron committed:
Change margins

Gargron committed:
Fix visual glitch with private mentions

renchap committed:
Support streaming notifications

renchap committed:
Use a constant for the max amount of avatars to display in a group

renchap committed:
Fix i18n

renchap committed:
Add back default avatar size

renchap committed:
Rename notificationAPI JSON types for consistency

renchap committed:
Handle removals from notification groups

renchap committed:
Fix import loop

renchap committed:
Handle unread notifications

renchap committed:
Fix duplicate accounts in a group

renchap committed:
Avoid having 2 gaps at the top of the notifications after a disconnect

renchap committed:
Fix i18n issue

ClearlyClaire committed:
Improve gap handling

ClearlyClaire committed:
Add slow mode support

ClearlyClaire committed:
Fix linting issue

renchap committed:
Add hover cards to grouped notifications

ClearlyClaire committed:
Add notifications store trimming

ClearlyClaire committed:
Add notifications grouping opt-in

vmstan commented:
In the current implementation there appears to be some flakiness in notifications being cleared out for post edit and poll ending notifications. Both scrolling the page or hitting the double checkmark button does not reset the counter. It clears out only on a manual refresh of the page after hitting the reset button.

<img width="923" alt="image" src="https://github.com/mastodon/mastodon/assets/3002053/81f5e912-7479-406a-8487-9c2179122a85">

vmstan commented:
Tooltips are still shown combination with the hover cards.
<img width="434" alt="image" src="https://github.com/mastodon/mastodon/assets/3002053/f6f08c38-3d20-4648-b74b-d742053da3d8">

renchap committed:
Load the initial notifications when the beta feature is enabled

ClearlyClaire committed:
Handle `fetchMarkers` action in markers reducer

ClearlyClaire committed:
Implement unread notification handling

renchap committed:
Add a link to view all the boosts and favorites from a grouped notification

renchap committed:
Re-initialize the notifications when opting in/out of the new feature

renchap committed:
Update label for private mentions

Gargron committed:
Change style of unread notifications

Gargron committed:
Change style of private mentions in notifications

ClearlyClaire committed:
Port “Move filtered notifications bar in scrollable area” to new component

ClearlyClaire committed:
Add accept/reject buttons on follow request notifications

ClearlyClaire committed:
Ensure graceful handling of grouped avatar overflow

ClearlyClaire committed:
Bump front-end grouped avatars limit to 8

ClearlyClaire committed:
Bump server-side sample accounts number to 8 for consistency with backend

ClearlyClaire committed:
Avoid N+1s when loading associated records for sample accounts

vmstan commented:
Follow back button on the notifications is a nice touch.

ClearlyClaire committed:
Fix N+1 on `accounts` within a notification group

ClearlyClaire committed:
Use database replica to load grouped notifications data

ClearlyClaire committed:
Small refactor

ClearlyClaire committed:
Add OTEL instrumentation for grouped notifications endpoint

ClearlyClaire committed:
Add OTel instrumentation for post and account formatting

ClearlyClaire committed:
Fix inconsistent overflow handling

ClearlyClaire committed:
Do not break up groups because of hours of inactivity

This decision was made under the assumption that group position would not be
bumped to the position of the newest notification. Unnecessarily breaking up
groups makes the grouping feel less useful.

ClearlyClaire committed:
Fix typos and inconsistent plurals in code

ClearlyClaire committed:
Rework OTel spans

ClearlyClaire committed:
Fix edge cases in last read notification handling

oneiros commented on app/controllers/api/v2_alpha/notifications_controller.rb:
```diff
@@ -12,10 +12,26 @@ def index
       @relationships = StatusRelationshipsPresenter.new(target_statuses_from_notifications, current_user&.account_id)
+
+      # Preload associations to avoid N+1s
+      ActiveRecord::Associations::Preloader.new(records: @grouped_notifications.flat_map(&:sample_accounts), associations: [:account_stat, { user: :role }]).call
```
Since `@grouped_notifications.flat_map(&:sample_accounts)` is now used in the OTEL instrumentation as well, it might make sense to cache this in an instance variable.

ClearlyClaire replied:
Ah, right, changed!

ClearlyClaire committed:
Address review remarks
